### PR TITLE
Scroll the search and filter dropdown

### PR DESCRIPTION
**Pull request by huwshimi**
## Done

- Constrain the search and filter dropdown to stay within the viewport.

### QA steps

- View the search and filter dropdown [default story in isolation](https://react-components-906.demos.haus/iframe.html?args=&id=search-and-filter--default-story&viewMode=story).
- Click on the input to open the dropdown.
- Resize your window so that the entire dropdown doesn't fit.
- The dropdown should resize so that it doesn't get cut off and the contents should scroll.

## Fixes

Fixes: https://github.com/canonical/jaas-dashboard/issues/707.

## Screenshots

Screen Recording 2023-03-20 at 4 27 18 pm

### CODE REVIEW

File: src/components/ContextualMenu/ContextualMenuDropdown/ContextualMenuDropdown.test.tsx [hunk 1]
```diff
@@ -17,15 +17,39 @@ describe("ContextualMenuDropdown ", () => {
   describe("adjustForWindow", () => {
     const generateFits = (overrides = {}) => {
       const fits = {
-        fromTop: { fitsAbove: true, fitsBelow: true },
-        fromBottom: { fitsAbove: true, fitsBelow: true },
-        fromLeft: { fitsLeft: true, fitsRight: true },
-        fromRight: { fitsLeft: true, fitsRight: true },
+        fromTop: {
```
Comment (huwshimi) [inline src/components/ContextualMenu/ContextualMenuDropdown/ContextualMenuDropdown.test.tsx:20]: Some of the tests mock the return values for the hook so they needed to be updated to include the new spacing values.

File: src/components/SearchAndFilter/SearchAndFilter.tsx [hunk 2]
```diff
@@ -177,6 +179,14 @@ const SearchAndFilter = ({
     };
   }, [searchData]);
 
+  useWindowFitment(
```
Comment (huwshimi) [inline src/components/SearchAndFilter/SearchAndFilter.tsx:182]: This listens to window resizes to get the coords for the dropdown.

File: src/hooks/useWindowFitment.ts [hunk 3]
```diff
@@ -134,24 +170,36 @@ export const useWindowFitment = (
           fromTop: {
             fitsAbove: referenceTop - heightIncludingSpace > windowTop,
             fitsBelow: referenceTop + heightIncludingSpace < windowBottom,
+            spaceAbove: Math.abs(windowTop - referenceTop),
```
Comment (huwshimi) [inline src/hooks/useWindowFitment.ts:173]: I've updated the return value to include the space available, not just whether the item can fit.

**Review by codeEmpress1 — APPROVED** [review]

**Comment by bartaz** [comment]
Should there be some min-height on the dropdown?

It works fine if search is on top of the window, but if it's close to the bottom (because window is small or there is content above it, the dropdown will be too small to use:

image

**Comment by huwshimi** [comment]
> Should there be some min-height on the dropdown?
> 

That's a good idea, I've set it to have a min height like this:

Screenshot 2023-03-21 at 9 11 14 am

**Review by ClementChaumel — APPROVED** [review]
Excellent 🎉

**Pull request merged**